File: app/src/main/java/pe/andy/bookholic/adapter/BookAdapter.java
package pe.andy.bookholic.adapter;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import java.util.List;

import lombok.Getter;
import lombok.Setter;
import pe.andy.bookholic.databinding.BookItemBinding;
import pe.andy.bookholic.model.Ebook;
import pe.andy.bookholic.util.BookColor;

public class BookAdapter extends RecyclerView.Adapter<BookAdapter.BookViewHolder> {

    private Context mContext;
    @Getter @Setter
    private List<Ebook> books;

    private BookColor bookColor;

    public BookAdapter(Context mContext, List<Ebook> books) {
        this.mContext = mContext;
        this.books = books;

        bookColor = new BookColor(mContext);
    }

    @Override
    @NonNull
    public BookViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        BookItemBinding binding = BookItemBinding.inflate(
                LayoutInflater.from(mContext), parent, false);

        return new BookViewHolder(binding.getRoot(), binding);
    }

    @Override
    public void onBindViewHolder(@NonNull BookViewHolder holder, int position) {
        Ebook book = this.books.get(position);

        BookItemBinding binding = holder.binding;

        binding.title.setText(book.getTitle());
        binding.author.setText(book.getAuthor());
        binding.publisher.setText(book.getPublisher());
        binding.date.setText(book.getDate());
        binding.platform.setText(book.getPlatform());
        binding.platform.setBackgroundColor(bookColor.getPlatformBGColor(book.getPlatform()));
        binding.libraryName.setText(book.getLibraryName());

        // Visibility
        int visibility = (book.getCountTotal() < 0) ? View.INVISIBLE : View.VISIBLE;
        binding.rentCount.setVisibility(visibility);

        // Color
        int color = bookColor.getRentStatusColor(book);
        binding.rentCount.setBackgroundColor(color);

        // Text
        String count = book.getCountRent() + " / " + book.getCountTotal();
        binding.rentCount.setText(count);

        // Bind onclick url
        binding.bookView.setOnClickListener(view -> {
            Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(book.getUrl()));
            mContext.startActivity(browserIntent);
        });

        try {
            Glide.with(this.mContext)
                    .load(book.getThumbnailUrl())
                    .into(binding.thumbnail);
        }
        catch (Exception ex) {
            Log.d("Bookholic", "Thumbnail url loading failure: " + book.getThumbnailUrl());
        }

    }

    @Override
    public int getItemCount() {
        return (this.books == null) ? 0 : this.books.size();
    }

    class BookViewHolder extends RecyclerView.ViewHolder {

        BookItemBinding binding;

        BookViewHolder(@NonNull View itemView, BookItemBinding binding) {
            super(itemView);
            this.binding = binding;
        }
    }

}

File: app/src/main/java/pe/andy/bookholic/adapter/BookAdapter.kt
package pe.andy.bookholic.adapter

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import coil.load
import pe.andy.bookholic.databinding.BookItemBinding
import pe.andy.bookholic.model.Ebook
import pe.andy.bookholic.util.BookColor
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

class BookAdapter : RecyclerView.Adapter<BookAdapter.BookViewHolder>() {

    var books = mutableListOf<Ebook>()

    lateinit var bookColor: BookColor

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BookViewHolder {

        val binding = BookItemBinding.inflate(
                LayoutInflater.from(parent.context), parent, false)

        bookColor = BookColor(parent.context)

        return BookViewHolder(binding)
    }

    override fun getItemCount(): Int
            = this.books.size

    override fun onBindViewHolder(holder: BookViewHolder, position: Int) {
        val book = this.books[position]
        holder.bind(book)
    }

    fun add(list: List<Ebook>) {
        books.addAll(list)
        notifyDataSetChanged()
    }

    fun set(list: List<Ebook>) {
        books.clear()
        books.addAll(list)
        notifyDataSetChanged()
    }

    fun clear() {
        books.clear()
        notifyDataSetChanged()
    }

    inner class BookViewHolder(val binding: BookItemBinding) : RecyclerView.ViewHolder(binding.root) {

        fun bind(book: Ebook) {

            with(binding) {
                title.text = book.title
                author.text = book.author
                publisher.text = book.publisher
                date.text = book.date
                platform.text = book.platform
                platform.setBackgroundColor(bookColor.getPlatformBGColor(book.platform))
                libraryName.text = book.libraryName

                rentCount.apply {
                    // Visibility
                    visibility = when {
                        book.countTotal < 0 -> View.INVISIBLE
                        else -> View.VISIBLE
                    }

                    // Color
                    setBackgroundColor(bookColor.getRentStatusColor(book))

                    // Text
                    @SuppressLint("SetTextI18n")
                    text = "${book.countRent} / ${book.countTotal}"
                }

                header.apply {
                    val backgroundColor = when {
                        book.isAvailable -> bookColor.bgGreen
                        book.countTotal < 0 -> bookColor.bgLightBlue
                        else -> bookColor.bgRed
                    }
                    setBackgroundColor(backgroundColor)
                }

                imageRecent.visibility = when {
                    isRecentlyPublished(book.date) -> View.VISIBLE
                    else -> View.GONE
                }

                // Bind onclick url
                bookView.setOnClickListener {
                    if (book.url.isNotEmpty()) {
                        val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(book.url))
                        bookView.context.startActivity(browserIntent)
                    }
                }

                thumbnail.load(book.thumbnailUrl)
            }
        }

        /**
         * 2년 내에 출판된 책인지 확인
         */
        internal fun isRecentlyPublished(date: String): Boolean {
            try {
                val dateTimeFormatter = when {
                    date.matches("""^\d\d\d\d-\d\d-\d\d""".toRegex()) -> DateTimeFormatter.ISO_DATE
                    date.matches("""^\d\d\d\d\d\d\d\d""".toRegex()) -> DateTimeFormatter.ofPattern("yyyyMMdd")
                    else -> return false
                }

                return ChronoUnit.DAYS.between(
                    LocalDate.parse(date, dateTimeFormatter),
                    LocalDate.now()) < (365 * 2)
            } catch (e: Exception) {
                e.printStackTrace()
                return false
            }
        }
    }
}
